http_aws_sigv4: skip special user header entries

jay opened the pull request:
- Do not include user header entries in the format 'name:' with no value in AWS SignedHeaders.

User header entries in that format are not actually used to specify a no-value header, instead they are used to signal that the user wants that internal header removed. For example if the user passes a header "Expect:" then that means they want libcurl to not send that header.

Prior to this change those header names were included in AWS SignedHeaders.

Reported-by: apparentorder@users.noreply.github.com

Fixes #11664
Closes #xxxx

---

Note I did not cover the case if the user actually does want a no-value header, which is specified as ending in a semicolon, eg `-H "foo;"` causes "foo:" header to be sent which means AWS SignedHeaders would show something like  `SignedHeaders=content-type;foo;;host;x-amz-date, Signature=...`. Someone familiar with AWS will need to see if that is correct behavior.

/cc @outscale-mgo

outscale-mgo commented:
I've reread this https://docs.aws.amazon.com/IAM/latest/UserGuide/create-signed-request.html

I think AWS tend to assume the header will be something like `"NAME: VALUE"`. and doesn't really specify what happen if VALUE is not present (I might have miss a part of the doc though).

So I guess `foo;;` is okay
But I think your code skip the header for the SignedHeader, but not for `Canonical header` computation here: 
https://github.com/curl/curl/pull/11668/files#diff-1d1fee636fd00dd0694ff6db673a8b5ab1ec1d3db540e748645f7b406cf64843R261
Which I guess should not include `'name:'` either. (maybe removing `name:` from the header list at the begin make_headers would be easier than spiking it every time)
Also a test would be nice :)
Thanks for improving signature v4 code.

bagder commented on lib/http_aws_sigv4.c:
```diff
@@ -200,6 +200,15 @@ static CURLcode make_headers(struct Curl_easy *data,
+       used to ignore an internal header so do not add it to the header list */
+    if(ptr) {
+      for(++ptr; ISSPACE(*ptr); ++ptr)
+        ;
```
I think we can insist on `name:` with no spaces at all when removing headers, so this check could be simplified a little.

jay commented:
So this turned out to be a bit more complicated than I thought. I redid the changes to handle both 'name:' and 'name;'. There is still a check for whitespace because I want to ignore value of whitespace only like `'name:   '` and `'name;   '` and missing separator 'name'  since those would not actually be sent. A weird quirk I noticed while testing this is the Host header if specified with only whitespace is sent by libcurl, unlike arbitrary headers. So for example
~~~
curl -v -H "foo;" -H "bar:" -H "baz:   " -H "Host:   " http://httpbin.org

> GET / HTTP/1.1
> Host:
> User-Agent: curl/8.3.0-DEV
> Accept: */*
> foo:
~~~
Since that is bad syntax I did not special case any further and the code ignores the value of whitespace only in all cases. Technically though if someone did a corner case like this they would get a bad signature, not that it would matter though because the request is already bad.

jay commented:
> But I think your code skip the header for the SignedHeader, but not for `Canonical header` computation here:

https://github.com/curl/curl/blob/c90c78333bbd15d171d26d3d4d005066ed152ca5/lib/http_aws_sigv4.c#L252-L270

true but it's iterating through the signed header list `head` (the just built list that skips the special user headers) and not through `data->set.headers` (the user headers unmodified)

> Also a test would be nice :)

I've updated and added comments to test 1955 about the special cases and I will need to recalculate the signature

bagder commented:
> Linux torture / torture Failing after 46m

^^
